File: src/go/dp/maxProfit.go
package dp

import "math"

// 121. 买卖股票的最佳时机
// 永远用当前点与当前点之前的最低点进行比较就可以得出最大的利润了
func maxProfit(prices []int) int {
	maxProfit := 0
	minPrices := math.MaxInt32
	for i := 0; i < len(prices); i++ {
		if minPrices > prices[i] {
			minPrices = prices[i]
		} else if prices[i]-minPrices > maxProfit {
			maxProfit = prices[i] - minPrices
		}
	}
	return maxProfit
}

File: feishu/MaxProfit.java
/**
 * @author Zexho
 * @date 2022/3/26 8:49 PM
 */
public class MaxProfit {

    public int maxProfit(int[] prices) {
        int min = prices[0], max = 0, pro;
        for (int i = 1; i < prices.length; i++) {
            max = (pro = prices[i] - min) > max ? pro : max;
            min = Math.min(min, prices[i]);
        }
        return max;
    }

}
